src/link-preview-expander.ts

```diff
@@ -1,16 +1,194 @@
 import * as sourcegraph from 'sourcegraph'
+import { checkIsURL, cleanURL, createMetadataCache, FAILURE, getWord } from './util'
+import parse from 'node-html-parser'
 
-export function activate(ctx: sourcegraph.ExtensionContext): void {
-    ctx.subscriptions.add(
+export function activate(context: sourcegraph.ExtensionContext): void {
+    const metadataCache = createMetadataCache({})
+
+    context.subscriptions.add(
         sourcegraph.languages.registerHoverProvider(['*'], {
-            provideHover: () => ({
-                contents: {
-                    value: 'Hello world from link-preview-expander! 🎉🎉🎉',
-                    kind: sourcegraph.MarkupKind.Markdown
+            provideHover: (document, position) => {
+                const range = document.getWordRangeAtPosition(position)
+                if (!range) {
+                    return null
+                }
+                const maybeURL = cleanURL(getWord(document, range))
+                const isURL = checkIsURL(maybeURL)
+
+                if (!isURL) {
+                    return null
+                }
+
+                /**
+                 * Creates the markdown string to be rendered in the hover tooltip.
+                 */
+                const createResult: (metadata?: Metadata) => sourcegraph.Badged<sourcegraph.Hover> = metadata => {
+                    const { image, title, description } = metadata ?? {}
+
+                    let markdownContent = `<h4><a href="${maybeURL}" target="_blank" rel="noopener noreferrer">${
+                        title || maybeURL
+                    }</a></h4>`
+
+                    if (image || description) {
+                        markdownContent += '<hr />'
+
+                        if (image) {
+                            markdownContent += `<img height="64" src="${image}" />`
```

Comment:
I'd float this to the left
```suggestion
                            markdownContent += `<img height="64" src="${image}" align="left" />`
```